superset/assets/src/datasource/ChangeDatasourceModal.jsx

```diff
@@ -0,0 +1,169 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import React from 'react';
+import PropTypes from 'prop-types';
+import { Table } from 'reactable-arc';
+import {
+  FormControl,
+  Modal,
+} from 'react-bootstrap';
+import { SupersetClient } from '@superset-ui/connection';
+import { t } from '@superset-ui/translation';
+
+import getClientErrorObject from '../utils/getClientErrorObject';
+import Loading from '../components/Loading';
+import withToasts from '../messageToasts/enhancers/withToasts';
+
+const propTypes = {
+  addDangerToast: PropTypes.func.isRequired,
+  onChange: PropTypes.func,
+  onDatasourceSave: PropTypes.func,
+  onHide: PropTypes.func,
+  show: PropTypes.bool.isRequired,
+};
+
+const defaultProps = {
+  onChange: () => {},
+  onDatasourceSave: () => {},
+  onHide: () => {},
+};
+
+class ChangeDatasourceModal extends React.PureComponent {
+  constructor(props) {
+    super(props);
+    this.state = {
+      loading: true,
+      datasources: null,
+    };
+    this.setSearchRef = this.setSearchRef.bind(this);
+    this.onEnterModal = this.onEnterModal.bind(this);
+    this.selectDatasource = this.selectDatasource.bind(this);
+    this.changeSearch = this.changeSearch.bind(this);
+  }
+
+  onEnterModal() {
+    if (this.searchRef) {
+      this.searchRef.focus();
+    }
+    const that = this;
+    if (!this.state.datasources) {
+      SupersetClient.get({
+        endpoint: '/superset/datasources/',
+      })
+        .then(({ json }) => {
+          const datasources = json.map(ds => ({
+            rawName: ds.name,
+            connection: ds.connection,
+            schema: ds.schema,
+            name: (
+              <a
+                href="#"
+                onClick={this.selectDatasource.bind(this, ds)}
+                className="datasource-link"
+              >
+                {ds.name}
+              </a>
+            ),
+            type: ds.type,
+          }));
+
+          that.setState({ loading: false, datasources });
+        })
+        .catch((response) => {
+          that.setState({ loading: false });
+          getClientErrorObject(response).then(({ error }) => {
+            this.props.addDangerToast(error.error || error.statusText || error);
+          });
+        });
+    }
+  }
+
+  setSearchRef(searchRef) {
+    this.searchRef = searchRef;
```

Comment:
Any reason you didn't just do this in the constructor vs. doing a function/bind?